ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/Send.hs

```diff
@@ -0,0 +1,252 @@
+  QBoxLayout.addWidget amountLayout unitLabel
+  QBoxLayout.setSpacing amountLayout 0
+  addRowLayout layout amountLabel amountLayout
+  addSeparator layout
+
+  fromAccountsLabel <- QLabel.newWithText ("<b>FROM</b>" :: String)
+  fromDisplay <- QLabel.newWithText ("select account..." :: String)
+  QObject.setObjectName fromDisplay ("accountsSelectorDisplay" :: String)
+
+  case sendInputs of
+    SendInputsSingle {} -> return ()
+    SendInputsMulti {} -> do
+      addRow layout fromAccountsLabel fromDisplay
+      addSeparator layout
+
+  addressLabel <- QLabel.newWithText ("<b>ADDRESS</b>" :: String)
+  addressEdit <- QLineEdit.new
+
+  addRow layout addressLabel addressEdit
+  addSeparator layout
+
+  sendButton <- QPushButton.newWithText ("SEND" :: String)
+  void $ setProperty sendButton ("dialogButtonRole" :: Text) ("dialogAction" :: Text)
+
+  QBoxLayout.addWidget layout sendButton
+
+  QBoxLayout.addStretch layout
+
+  accountSelector <- case sendInputs of
+    SendInputsSingle aIdx -> return $ Left aIdx
+    SendInputsMulti uacis -> do
+      selector@AccountSelector{..} <- createAccountSelector uacis
+      QWidget.setParent dropdownWidget send
+      QWidget.hide dropdownWidget
+
+      QWidget.setMinimumWidth fromDisplay =<< QWidget.width dropdownWidget
+
+      return $ Right selector
+
+  let s = Send{..}
+
+  connect_ sendButton QAbstractButton.clickedSignal $ \_ -> QDialog.accept send
+  connect_ amountEdit QLineEdit.textChangedSignal $ \_ -> revalidate s
+  connect_ addressEdit QLineEdit.textChangedSignal $ \_ -> revalidate s
+
+  whenRight accountSelector $ \as@AccountSelector{..} -> do
+    void $ Event.onEvent fromDisplay $
+      \(ev :: QMouseEvent.QMouseEvent) -> do
+        eventType <- QEvent.eventType ev
+        if eventType == QEvent.MouseButtonRelease
+          then toggleDropDown dropdownWidget s $> True
+          else return False
+
+    void $ Event.onEvent send $
+      \(ev :: QEvent.QEvent) -> do
+        eventType <- QEvent.eventType ev
+        dropdownVisible <- QWidget.isVisible dropdownWidget
+        if eventType == QEvent.Resize && dropdownVisible
+          then moveDropDown dropdownWidget s $> True
+          else return False
+
+    for_ checkboxes $
+      \AccountCheckbox{..} -> connect_ checkbox QAbstractButton.toggledSignal $ \_ -> accountsSelected as s
+
+  revalidate s
+
+  return s
+
+runSend :: SendInputs -> IO SendResult
+runSend sendInputs = do
+  s@Send{send = send} <- initSend sendInputs
+  result <- toEnum <$> QDialog.exec send
+
+  options <- fillSendOptions s
+
+  return $ case result of
+    QDialog.Accepted -> case options of
+      Just so -> SendSuccess so
+      Nothing -> SendCancel
+    QDialog.Rejected -> SendCancel
+
+fillSendOptions :: Send -> IO (Maybe SendOptions)
+fillSendOptions Send{..} = do
+  amount <- fromString <$> QLineEdit.text amountEdit
+  address <- fromString <$> QLineEdit.text addressEdit
+  -- TODO do we want to validate any of this here? ^
```

Comment:
I'd prefer to validate address format here, but I'm not sure how to do it properly.
Also, amount should probably be validated, since Cardano has precision only up to 10^-6, for example.

Reply:
> I'd prefer to validate address format here, but I'm not sure how to do it properly.

Do you mean "which Cardano function to call" or "how to expose this function to UI"? In the former case it's `decodeTextAddress`, in the latter it should be exposed via knit somehow or, it appears to be too cumbersome, at least similar to `generateMnemonic`.

> Also, amount should probably be validated, since Cardano has precision only up to 10^-6, for example.

There is `Pos.Core.Common.Coin.checkCoin`, but it works with integer. Backend doesn't know about 10⁻⁶ precision, it should be specified somewhere in Glue.

Reply:
I remembered why I did not want to make `generateMnemonic` work through knit -- it disrupts the workflow: i have to send Knit comand and wait for the result _asynchronously_. That means that validation is no longer simple `Send -> IO Bool` function, it gets scattered into two places at least, and some state has to be stored somewhere.

I had an idea of making `KnitT` transformer that will hide this in its `>>=`, but probably nobody got time for that ATM.

So I'd go with `generateMnemonic`-like solution.

This has to be in Glue anyway.

CC @goodsoft.

Reply:
This comment is now obsolete, isn't it?

Reply:
Right, will drop